OpenRA.Mods.Common/Traits/TargetableUnit.cs

```diff
@@ -20,20 +21,30 @@ public class TargetableUnitInfo : ITraitInfo, ITargetableInfo
 		public readonly string[] TargetTypes = { };
 		public string[] GetTargetTypes() { return TargetTypes; }
 
+		[Desc("Upgrades that, when granted, trigger target type modification.")]
+		public readonly string[] UpgradeTypes = { };
+
+		[Desc("List of targets to add or remove when at least one upgrade is granted. ",
+			"Prefix target types to remove with an exclamation point (!). ",
+			"Use the exclamation point alone to remove all target types.")]
```

Comment:
We tend to avoid magic values where possible, so lets make this a full list of types instead, please.  You could default this to null, and then have it fall back to TargetTypes in that case.